src/promptflow/promptflow/_sdk/_service/apis/telemetry.py

```diff
@@ -0,0 +1,148 @@
+
+
+def _dict_camel_to_snake(data):
+    if isinstance(data, dict):
+        result = {}
+        for key, value in data.items():
+            result[camel_to_snake(key)] = _dict_camel_to_snake(value)
+        return result
+    else:
+        return data
+
+
+def parse_activity_info(telemetry_data, user_agent):
+    from promptflow._sdk._telemetry.activity import generate_request_id
+
+    first_call = telemetry_data.get("firstCall", True)
+    request_id = telemetry_data.get("requestId", generate_request_id())
+
+    return {
+        "request_id": request_id,
+        "first_call": first_call,
+        "user_agent": user_agent,
+        **_dict_camel_to_snake(telemetry_data.get("metadata", {})),
+    }
+
+
+def validate_telemetry_payload(payload):
+    from promptflow._sdk._telemetry import ActivityCompletionStatus, ActivityType
+
+    if not all(key in payload for key in ("eventType", "timestamp", "metadata")):
+        missing_fields = {"eventType", "timestamp", "metadata"} - set(payload.keys())
+        raise UserErrorException(f"Missing required fields in telemetry payload: {', '.join(missing_fields)}")
+    if payload.get("eventType") not in (EventType.START, EventType.END):
+        raise UserErrorException(f"Invalid eventType: {payload.get('eventType')}")
+    metadata = payload.get("metadata")
+    if not isinstance(metadata, dict):
+        raise UserErrorException(f"Invalid metadata: {metadata}")
+    if metadata.get("activityName") not in [
+        AllowedActivityName.FLOW_TEST,
+        AllowedActivityName.FLOW_NODE_TEST,
+        AllowedActivityName.GENERATE_TOOL_META,
+    ]:
+        raise UserErrorException(f"Invalid activityName: {metadata.get('activityName')}")
+    if metadata.get("activityType") not in [
+        ActivityType.INTERNALCALL,
+        ActivityType.PUBLICAPI,
+    ]:
+        raise UserErrorException(f"Invalid activityType: {metadata.get('activityType')}")
+
+    if payload.get("eventType") == EventType.END:
+        if not all(
+            key in metadata
+            for key in (
+                "requestId",  # End event should have requestId returned from start event
+                "completionStatus",  # End event should have completionStatus
+                "durationMs",  # End event should have durationMs
+            )
+        ):
+            missing_fields = {"requestId", "completionStatus", "durationMs"} - set(metadata.keys())
+            raise UserErrorException(f"Missing required fields in telemetry metadata: {', '.join(missing_fields)}")
+
+        if metadata.get("completionStatus") not in [
+            ActivityCompletionStatus.FAILURE,
+            ActivityCompletionStatus.SUCCESS,
+        ]:
+            raise UserErrorException(f"Invalid completionStatus: {metadata.get('completionStatus')}")
+
+        if metadata.get("completionStatus") == ActivityCompletionStatus.FAILURE:
+            if not all(
+                key in metadata
+                for key in (
```

Comment:
Are these fields not required? (errorCategory, ...)
If the user uses an old version, then these fields will not exist.

Reply:
This is an API to record telemetry and this a new API, so C# can always provide them